Composer/packages/extension/src/utils/npm.ts

```diff
@@ -1,65 +1,81 @@
+import { promisify } from 'util';
+
+import { mkdir } from 'fs-extra';
+import fetch from 'node-fetch';
+import tar from 'tar';
 
 import logger from '../logger';
+import { ExtensionSearchResult } from '../types/extension';
+
+const streamPipeline = promisify(require('stream').pipeline);
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires, import/order
+const rimraf = promisify(require('rimraf'));
 
 const log = logger.extend('npm');
 
-type NpmOutput = {
-  stdout: string;
-  stderr: string;
-  code: number;
-};
-type NpmCommand = 'install' | 'uninstall' | 'search' | 'link';
-type NpmOptions = {
-  [key: string]: string;
-};
-
-function processOptions(opts: NpmOptions) {
-  return Object.entries({ '--no-fund': '', '--no-audit': '', '--quiet': '', ...opts }).map(([flag, value]) => {
-    return value ? `${flag}=${value}` : flag;
-  });
-}
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export async function search(query = ''): Promise<ExtensionSearchResult[]> {
+  try {
+    log('Searching for %s', query);
+    const queryString = query.replace(' ', '+');
+    const res = await fetch(
+      `https://registry.npmjs.org/-/v1/search?text=${queryString}+keywords:botframework-composer&size=100&from=0&quality=0.65&popularity=0.98&maintenance=0.5`
+    );
+    const data = await res.json();
 
-/**
- * Executes npm commands that include user input safely
- * @param `command` npm command to execute.
- * @param `args` cli arguments
- * @param `opts` cli flags
- * @param `spawnOpts` options to pass to spawn command
- * @returns Object with stdout, stderr, and exit code from command
- */
-export async function npm(
-  command: NpmCommand,
-  args: string,
-  opts: NpmOptions = {},
-  spawnOpts: SpawnOptionsWithoutStdio = {}
-): Promise<NpmOutput> {
-  return new Promise((resolve, reject) => {
-    const cmdOptions = processOptions(opts);
-    const spawnArgs = [command, ...cmdOptions, args];
-    log('npm %s', spawnArgs.join(' '));
-    let stdout = '';
-    let stderr = '';
-
-    const proc = spawn('npm', spawnArgs, { ...spawnOpts, shell: process.platform === 'win32' });
-
-    proc.stdout.on('data', (data) => {
-      stdout += data;
-    });
+    log('Got %d result(s).', data.objects?.length ?? 0);
 
-    proc.stderr.on('data', (data) => {
-      stderr += data;
-    });
+    return data.objects.map((result) => {
+      const { name, version, description = '', keywords = [], links = {} } = result.package;
 
-    proc.on('close', (code) => {
-      if (code > 0) {
-        reject({ stdout, stderr, code });
-      } else {
-        resolve({ stdout, stderr, code });
-      }
+      return {
+        id: name,
+        version,
+        description,
+        keywords,
+        url: links.npm ?? '',
+      } as ExtensionSearchResult;
     });
+  } catch (err) {
+    log('%O', err);
+
+    return [];
+  }
+}
+
+export async function downloadPackage(name: string, versionOrTag: string, destination: string) {
```

Comment:
try / catch error handling?

Reply:
I think I'm ok having the caller handle errors for this one. What do you think?

Reply:
That's fine by me. Looks like its only invocation right now is wrapped in a try 👍